### PR TITLE
RegionCode class added with one hierarchy attribute

### LINKED ISSUE #194: Add RegionCode class
For improved clarity it would be nice to add a `RegionCode` which has the additional attribute hierarchy.

### PR DESCRIPTION
**Pull request by GretchenSchowalter**
I just created a simple RegionCode class in code.py and updated the RegionCodeList class. Let me know if I missed something or did this incorrectly. Closes #194

### CODE REVIEW

File: nomenclature/code.py [hunk 1]
```diff
@@ -194,3 +194,18 @@ def agg_kwargs(self) -> Dict[str, Any]:
...
+    Attributes
+    ----------
+    name : str
+        Name of the RegionCode
+    hierarchy : str
+        Hierarchy of the RegionCode
+
+    """
+
+    hierarchy: Optional[str] = None
```
Comment (danielhuppmann) [inline nomenclature/code.py:211]: This should be a required attribute.
```suggestion
    hierarchy: str
```

File: nomenclature/codelist.py [hunk 2]
```diff
@@ -548,7 +548,7 @@ def from_directory(cls, name: str, path: Path, file_glob_pattern: str = "**/*"):
         RegionCodeList
 
         """
-        code_list: List[Code] = []
+        code_list: List[RegionCode] = []
```
Comment (phackstock) [inline nomenclature/codelist.py:551]: Just for completeness I would add this line above:
```python
mapping: Dict[str, RegionCode] = {}
```
and this one below:
```python
code_basis: ClassVar = RegionCode
```

File: nomenclature/codelist.py [hunk 3]
```diff
@@ -562,14 +562,14 @@ def from_directory(cls, name: str, path: Path, file_glob_pattern: str = "**/*"):
             for top_level_cat in _code_list:
                 for top_key, _codes in top_level_cat.items():
                     for item in _codes:
-                        code = Code.from_dict(item)
+                        code = RegionCode.from_dict(item)
```
Comment (phackstock) [inline nomenclature/codelist.py:565]: You could streamline this slightly: 
```suggestion
                        code_list.append(RegionCode.from_dict({
                        **item,
                         "hierarchy": top_key,
                          "file":yaml_file.relative_to(path.parent).as_posix()})
                          )
```
although I'm not sure it's much more readable.

**Review by phackstock — COMMENTED** [review]
Two small suggestions and a comment other than that it looks good to me.

File: nomenclature/codelist.py (hunk 3 above)
Comment (phackstock) [reply nomenclature/codelist.py:565]: As discussed bilaterally, please disregard.

**Review by danielhuppmann — COMMENTED** [review]
I'm wondering why there is still a dedicated method `filter()` in the **RegionCodeList** class - wouldn't this now be covered by the inherited method of the **CodeList** class?

**Comment by GretchenSchowalter** [comment]
Yep, that makes sense. I will take the method and the corresponding tests out.

**Comment by danielhuppmann** [comment]
No, leave the tests! They should continue to work by automatically inheriting the `filter()` method from the **CodeList** class.

**Comment by phackstock** [comment]
Also taking it out without changing anything about `CodeList.filter` would not be good because it return a generic `CodeList`.
Then we would have the strange situation in `RegionCodeList.filter` would return a `CodeList` and not a `RegionCodeList`.
You'd have to adapt `CodeList.filter` to return whatever CodeList it was called from, i.e. `CodeList` for a generic one and `RegionCodeList` for a RegionCodeList.

**Comment by danielhuppmann** [comment]
Sorry, I think that my comment sent you down a rabbit hole that we don't need to open right now. Can you simply re-insert the `RegionCodeList.filter()` and the corresponding tests? Then the PR is good to be merged - and we can worry about proper sub-classing later...

**Review by phackstock — APPROVED** [review]
Looks good to me as well.

**Pull request merged**